File: two-sum.ts
function twoSum(nums: number[], target: number): number[] {
  const mem = {};
  
  const output = [];
  for(let idx = 0; idx < nums.length; idx++) {
      
      if ( nums[idx] in mem ) {
          output.push(mem[nums[idx]], idx)
      } else {
          mem[target - nums[idx]] = idx;
      }
      
  }
  return output;
};

function twoSumAgain(nums: number[], target: number): number[] {
    const map = {};
    
    for (let i = 0; i< nums.length;i++) {
        const curr = nums[i]
        const toSearch = target - curr;
        
        if ( toSearch in map ) {
            return [i, map[toSearch]]
        }
        
        map[curr] = i;
        
    }

    return [-1, -1]
};

File: restart_prep/two_sum.cpp
#include <cstddef>
#include <vector>
#include <unordered_map>

class Solution {
public:
	std::vector<int> twoSum(std::vector<int>& nums, int target) {
		std::vector<int> res;
        std::unordered_map<int, int> seenBefore;
        for (int i = 0; i < nums.size(); i++) {
            int curr = nums[i];
			int check = target - curr;
			std::unordered_map<int, int>::iterator seenBeforeIterator = seenBefore.find(check);
			if (seenBeforeIterator != seenBefore.end()) {
				res.push_back(seenBeforeIterator->second);
				res.push_back(i);
				return res;
			}
			seenBefore.insert(std::make_pair(curr, i));
        }
		return res;
    }
};
